memtable/write_buffer_manager.cc

```diff
@@ -145,4 +178,38 @@ void WriteBufferManager::FreeMemWithCache(size_t mem) {
   (void)mem;
 #endif  // ROCKSDB_LITE
 }
+
+void WriteBufferManager::BeginWriteStall(StallInterface* wbm_stall) {
+  if (wbm_stall && ShouldStall()) {
+    std::unique_lock<std::mutex> lock(mu_);
+    queue_.push_back(wbm_stall);
+  }
+}
+
+// Called when memory is freed in FreeMem.
+void WriteBufferManager::EndWriteStall() {
+  if (enabled()) {
+    {
+      std::unique_lock<std::mutex> lock(mu_);
+      if (queue_.empty()) {
```

Comment:
Shouldn't ```IsStallThresholdExceeded``` be checked and ```stall_active_``` cleared anyway, even if the queue is empty? The calls to ```ShouldStall``` and ```BeginWriteStall``` are not atomic.